### ISSUE TITLE
Confusion about grad and loss computation in distributed multi-gpu training

### ISSUE THREAD

**Issue by Galaxy-Husky** [OpenNMT/OpenNMT-py#2451]
From the discussion in #2022, I have figured out that the loss is accumulated over N GPUs and divided by the normalization (the total batch size). But I'm confused why we need do one more gradient all_reduce as followed.

https://github.com/OpenNMT/OpenNMT-py/blob/4958bb188e131a35df965d26761f58e37db4d51d/onmt/trainer.py#L532-L540

Take the example from #2022. Let's say there are 2 gpus. The losses and batch sizes are loss1, b1 and loss2, b2, respectively.  And we compute $\frac{loss1}{b1 + b2}$ and $\frac{loss2}{b1 + b2}$, respectively. Is the above code snippet used to sum them to get $\frac{loss1 + loss2}{b1 + b2}$? If so, why do we need to rescale the summation by the number of gpus? 

I would appreciate it if someone could tell me the reason.

**Comment by vince62s** [comment]
If I recall properly (bear in mind it's been some time)
You are correct for the loss calculation and the loss calculation is correct.
This part above only rescale the gradients themselves because the all_reduc op is SUM so I think I just wanted to average the gradients between GPUs.

**Comment by canvas7** [comment]
Thanks, @vince62s. 
I'd like to be clear about what the target loss is. If our target is $\frac{loss1 + loss2}{b1 + b2}$, the SUM all_reduce op is enough to add $\frac{loss1}{b1 + b2}$ (on gpu1) and $\frac{loss2}{b1 + b2}$ (on gpu2).  If we rescale the gradients, will the target loss be $\frac{1}{2} * \frac{loss1 + loss2}{b1 + b2}$ ?

**Comment by Galaxy-Husky** [comment]
I still doubt that this rescaling operation will affect the calculation of loss. If you recall the reason, please let me know. Thank you for your reply.

**Comment by Galaxy-Husky** [comment]
Hi, @vince62s .

Since the distributed.py mentions that the code was heavily inspired by the equivalent of fairseq, I read the fairseq code. I found that they do the similar reduce and rescale. But they don't divide the loss by normalization (or sample_size in their code) before and they multiply the grads after that.

Let me take an example. Again, let's say we now have 2 gpus. 
First, in the code https://github.com/facebookresearch/fairseq/blob/100cd91db19bb27277a06a25eb4154c805b10189/fairseq/trainer.py#L910-L919, the _aggregate_logging_outputs func reduce_sum the sample_size to get $(b1 + b2)$ on each gpu.
Then, In the code https://github.com/facebookresearch/fairseq/blob/100cd91db19bb27277a06a25eb4154c805b10189/fairseq/trainer.py#L927-L930 , the all_reduce_grads func normalizes the grads by world_size to get $\frac{1}{2} *(loss1 + loss2)$ on each gpu.
Finally, in the code https://github.com/facebookresearch/fairseq/blob/100cd91db19bb27277a06a25eb4154c805b10189/fairseq/trainer.py#L946, the multiply_grads func multiplies the grads by (world_size / sample_size) to get $\frac{1}{2} *(loss1 + loss2) * \frac{2}{b1 + b2} = \frac{loss1 + loss2}{b1 + b2} $, which is our target exactly.

So I think the final result here is probably half the target as I said before.

**Comment by vince62s** [comment]
Again you are mixing up loss and gradients. The loss is correct.
I am not saying the code is 100% "standard" for the gradients compute but the only impact the current code could have vs a "standard" way is that the effective learning rate is halved. But the corss entropy value is correct.

Here are 3 runs:
2GPU - accum 4 - current code
[2023-07-26 08:44:22,538 INFO] Step 2000/10000; acc: 84.8; ppl:   7.3; xent: 2.0; lr: 0.00049; sents:  197961; bsz: 7647/9358/247; 53639/65636 tok/s;   2319 sec;
[2023-07-26 08:45:19,821 INFO] UPDATING VALIDATION BLEU
[2023-07-26 08:45:20,314 INFO] validation BLEU: 20.63400266532767
[2023-07-26 08:45:20,316 INFO] Train perplexity: 34.0465
[2023-07-26 08:45:20,316 INFO] Train accuracy: 62.4489
[2023-07-26 08:45:20,316 INFO] Sentences processed: 1.98207e+06
[2023-07-26 08:45:20,316 INFO] Average bsz: 7640/9357/248
[2023-07-26 08:45:20,316 INFO] Validation perplexity: 22.4512
[2023-07-26 08:45:20,316 INFO] Validation accuracy: 66.31
[2023-07-26 08:47:14,135 INFO] Step 2100/10000; acc: 85.6; ppl:   7.0; xent: 1.9; lr: 0.00052; sents:  206352; bsz: 7642/9355/258; 35627/43612 tok/s;   2490 sec;

1GPU - accum 8 - current code
[2023-07-26 09:24:52,836 INFO] Step 2000/10000; acc: 85.0; ppl:   7.2; xent: 2.0; lr: 0.00049; sents:  195600; bsz: 7634/9349/244; 60004/73487 tok/s;   2078 sec;
[2023-07-26 09:25:48,816 INFO] UPDATING VALIDATION BLEU
[2023-07-26 09:25:49,290 INFO] validation BLEU: 20.660578854627136
[2023-07-26 09:25:49,292 INFO] Train perplexity: 34.6823
[2023-07-26 09:25:49,292 INFO] Train accuracy: 62.0169
[2023-07-26 09:25:49,292 INFO] Sentences processed: 3.94453e+06
[2023-07-26 09:25:49,292 INFO] Average bsz: 7637/9351/247
[2023-07-26 09:25:49,292 INFO] Validation perplexity: 22.5795
[2023-07-26 09:25:49,292 INFO] Validation accuracy: 66.174
[2023-07-26 09:27:31,898 INFO] Step 2100/10000; acc: 85.0; ppl:   7.2; xent: 2.0; lr: 0.00052; sents:  192623; bsz: 7645/9361/241; 38451/47082 tok/s;   2237 sec;


2GPU - accum 4 - no rescaling of Grads (your suggestion "legacy" code)
[2023-07-26 10:08:24,272 INFO] Step 2000/10000; acc: 84.8; ppl:   7.3; xent: 2.0; lr: 0.00049; sents:  197961; bsz: 7647/9358/247; 53825/65864 tok/s;   2313 sec;
[2023-07-26 10:09:20,563 INFO] UPDATING VALIDATION BLEU
[2023-07-26 10:09:21,060 INFO] validation BLEU: 20.745329097688984
[2023-07-26 10:09:21,062 INFO] Train perplexity: 33.7705
[2023-07-26 10:09:21,062 INFO] Train accuracy: 62.589
[2023-07-26 10:09:21,062 INFO] Sentences processed: 1.98207e+06
[2023-07-26 10:09:21,062 INFO] Average bsz: 7640/9357/248
[2023-07-26 10:09:21,062 INFO] Validation perplexity: 22.4869
[2023-07-26 10:09:21,062 INFO] Validation accuracy: 66.2658
[2023-07-26 10:11:15,130 INFO] Step 2100/10000; acc: 85.7; ppl:   7.0; xent: 1.9; lr: 0.00052; sents:  206352; bsz: 7642/9355/258; 35781/43801 tok/s;   2484 sec;

Gradients are not proportional to the loss, they only give a direction to where to go.

**Comment by Galaxy-Husky** [comment]
@vince62s Thank you very much for your efforts to reply and run experiments!!

I mixed gradients and loss because I think gradients are proportional to the loss.  I don't want the loss to appear in one formula and the gradients the next. The goal of calculating the loss is to propagate the gradients.
Let's say $grad=\frac{d(loss)}{dx}$. If we double the loss, the new gradient will be doubled like $new\\_grad=\frac{d(2*loss)}{dx}=2 * \frac{d(loss)}{dx}=2 * grad$.  Could you tell me what goes wrong?

May I ask what optimizer you used in the 3 runs? If it was Adam, I don't think the result will be influenced too much by the bug, if it do exist.

**Comment by vince62s** [comment]
This is my last answer because keep repeating the same thing:
Gradients and Loss are proprotional ONLY in simple cases (eg: Linear function). In our use case, with multiple layer and cross-entropy calculation, there is no linearity between Gradiens and loss, hence your reasoning is incorrect.
